src/libsyntax/parse/parser.rs

```diff
@@ -2079,8 +2079,14 @@ impl<'a> Parser<'a> {
         let minus_present = self.eat(&token::BinOp(token::Minus));
         let lo = self.span;
         let literal = self.parse_lit()?;
-        let hi = self.prev_span;
-        let expr = self.mk_expr(lo.to(hi), ExprKind::Lit(literal), ThinVec::new());
+        // We don't store a span for the literal because it would be the same
+        // as the span for the expression. (Check this.)
+        debug_assert_eq!(literal.span,
+                         { let hi = self.prev_span;
+                           let expr_span = lo.to(hi);
+                           expr_span
+                         });
```

Comment:
This looks a bit suspicious.
Have you checked whether this assertion passes on literals passed to macros via `literal` and `expr` matchers?

Reply:
For
```rust
macro m { ... let x = $literal + y; ... }

m!(10);
```
I'd expect the expression span to point to `$literal`, but the literal span to point to `10`.

Reply:
For expressions (and other fragments) passed through multiple matchers we it would be nice to keep all the "backtrace" of spans by wrapping fragments into "invisible parentheses" on each substitution, but this isn't generally done now.

Reply:
I ran the tests and they passed. (In contrast, I tried doing the same no-span-for-literals transformation for HIR and the tests failed; that's how I found out about HIR's lack of parentheses being a problem.)

More compellingly, if you look at the code you can see that I haven't changed behaviour at all: `parse_lit` uses `self.span.to(self.prev_span)` for the span, where `self.span` is obtained at the start of `parse_lit` and `self.prev_span` is obtained at the end. And the new code does the same.

Reply:
>More compellingly, if you look at the code you can see that I haven't changed behaviour at all

OK.
This still goes into the wrong direction, because the span should ideally be different even if they are not right now.
I'll leave this to eddyb to decide, if AST is going to be radically rewritten for incremental parsing/expansion, then details like this won't matter anyway.